GeneratorInterface/Core/plugins/GenXSecAnalyzer.cc

```diff
@@ -1,29 +1,144 @@
 #include <iostream>
 #include <iomanip>
 
-GenXSecAnalyzer::GenXSecAnalyzer(const edm::ParameterSet& iConfig)
+// analyzer of a summary information product on filter efficiency for a user specified path
+// meant for the generator filter efficiency calculation
+
+// system include files
+#include <memory>
+#include <vector>
+#include <map>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/global/EDAnalyzer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Run.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/LuminosityBlock.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/thread_safety_macros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "SimDataFormats/GeneratorProducts/interface/GenLumiInfoProduct.h"
+#include "SimDataFormats/GeneratorProducts/interface/GenFilterInfo.h"
+#include "SimDataFormats/GeneratorProducts/interface/LHERunInfoProduct.h"
+
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+//
+// class declaration
+//
+namespace gxsec {
+  struct LumiCache {};
+  struct RunCache {
+    RunCache()
+        : product_(-9999),
+          filterOnlyEffRun_(0, 0, 0, 0, 0., 0., 0., 0.),
+          hepMCFilterEffRun_(0, 0, 0, 0, 0., 0., 0., 0.) {}
+    // for weight before GenFilter and HepMCFilter and before matching
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable double thisRunWeightPre_ = 0;
+
+    // for weight after GenFilter and HepMCFilter and after matching
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable double thisRunWeight_ = 0;
+
+    // GenLumiInfo before HepMCFilter and GenFilter, this is used
+    // for computation
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable GenLumiInfoProduct product_;
+
+    // statistics from additional generator filter, for computation
+    // reset for each run
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable GenFilterInfo filterOnlyEffRun_;
+
+    // statistics from HepMC filter, for computation
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable GenFilterInfo hepMCFilterEffRun_;
+
+    // the following vectors all have the same size
+    // LHE or Pythia/Herwig cross section of previous luminosity block
+    // vector size = number of processes, used for computation
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable std::map<int, GenLumiInfoProduct::XSec> previousLumiBlockLHEXSec_;
+
+    // LHE or Pythia/Herwig combined cross section of current luminosity block
+    // updated for each luminosity block, initialized in every run
+    // used for computation
+    CMS_THREAD_GUARD(GenXSecAnalyzer::mutex_) mutable std::map<int, GenLumiInfoProduct::XSec> currentLumiBlockLHEXSec_;
+  };
+}  // namespace gxsec
+
+class GenXSecAnalyzer
+    : public edm::global::EDAnalyzer<edm::RunCache<gxsec::RunCache>, edm::LuminosityBlockCache<gxsec::LumiCache>> {
```

Comment:
In https://github.com/cms-sw/cmssw/issues/29098 is was essentially concluded that this module does not have to be run in RelVals or in production. Now that the module is made thread-efficient, the issue becomes moot and I'll close it.